### PR TITLE
[LBSE] Fix clip-rule in masks

### PR DESCRIPTION
**Pull request by rwlbuis**
#### 018e5e8ed613ac7fd514ddbbf2cf7758b8ef2cca
<pre>
[LBSE] Fix clip-rule in masks
<a href="https://bugs.webkit.org/show_bug.cgi?id=265183">https://bugs.webkit.org/show_bug.cgi?id=265183</a>

Reviewed by Nikolas Zimmermann.

When in painting masks mode we should take clip-rule into account.

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* Source/WebCore/rendering/svg/RenderSVGResourceSolidColor.cpp:
(WebCore::RenderSVGResourceSolidColor::applyResource):

Canonical link: <a href="https://commits.webkit.org/271053@main">https://commits.webkit.org/271053@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/2bed7f94402e7f887b1d22dbe92f5befcafb9188

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27193 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5832 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28441 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29416 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24875 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7725 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3231 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24716 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27455 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4623 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23332 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/4038 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/4142 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24329 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30055 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24824 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24743 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30328 "") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4178 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/2331 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/28247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5644 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6552 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4637 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4558 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/rendering/svg/RenderSVGResourceSolidColor.cpp [hunk 1]
```diff
@@ -50,7 +50,9 @@ bool RenderSVGResourceSolidColor::applyResource(RenderElement& renderer, const R
         else
             context->setAlpha(1);
         context->setFillColor(style.colorByApplyingColorFilter(m_color));
-        if (!isRenderingMask)
+        if (isRenderingMask)
```
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/RenderSVGResourceSolidColor.cpp]: Ok. How does it work for legacy?

Comment (rwlbuis) [reply Source/WebCore/rendering/svg/RenderSVGResourceSolidColor.cpp]: It seems to have no effect there. I can add LBSE ifdef and runtime flag check if you want, looks like it is not needed though.

Comment (nikolaszimmermann) [reply Source/WebCore/rendering/svg/RenderSVGResourceSolidColor.cpp]: I rather meant to ask why this is not necessary for legacy?

**Comment by nikolaszimmermann** [comment]
Ah, `LegacyRenderSVGResourceClipper::drawContentIntoMaskImage` grabs the clip-rule from the style, and honors it:

```
        WindRule newClipRule = style.svgStyle().clipRule();
        bool isUseElement = child.hasTagName(SVGNames::useTag);
        if (isUseElement) {
            SVGUseElement& useElement = downcast<SVGUseElement>(child);
            renderer = useElement.rendererClipChild();
            if (!renderer)
                continue;
            if (!useElement.hasAttributeWithoutSynchronization(SVGNames::clip_ruleAttr))
                newClipRule = renderer->style().svgStyle().clipRule();
        }
```

**Comment by nikolaszimmermann** [comment]
So I think the idea is to not mutate the fill-rule during painting of the clipPath children -- instead propagate the correct clip-rule once in `RenderSVGResourceClipper`, before diving into `paintSVGResourceLayer()`. The legacy engine handled it this way, and it makes sense IMHO.

**Comment by rwlbuis** [comment]
> So I think the idea is to not mutate the fill-rule during painting of the clipPath children -- instead propagate the correct clip-rule once in `RenderSVGResourceClipper`, before diving into `paintSVGResourceLayer()`. The legacy engine handled it this way, and it makes sense IMHO.

I am not sure I agree, AFAICS the legacy engine sets the clip rule to be the fill rule per *child*, not set the clip-rule once. Seems like the PR does the same thing.

**Comment by nikolaszimmermann** [comment]
> > So I think the idea is to not mutate the fill-rule during painting of the clipPath children -- instead propagate the correct clip-rule once in `RenderSVGResourceClipper`, before diving into `paintSVGResourceLayer()`. The legacy engine handled it this way, and it makes sense IMHO.
> 
> I am not sure I agree, AFAICS the legacy engine sets the clip rule to be the fill rule per _child_, not set the clip-rule once. Seems like the PR does the same thing.

You are right, I was confused.

**Review by nikolaszimmermann — APPROVED** [review]
r=me

**Pull request merged**